### PR TITLE
TensorFlow Transformer Part-4

### PR DESCRIPTION
**Pull request by phi-dbq**
The implementation of `TFTransformer` based on previous steps.

### CODE REVIEW

File: python/sparkdl/graph/input.py [hunk 1]
```diff
@@ -92,6 +91,50 @@ def __init__(self, graph_def, input_tensor_name_from_signature,
         self.input_tensor_name_from_signature = input_tensor_name_from_signature
         self.output_tensor_name_from_signature = output_tensor_name_from_signature
 
+    def translateInputMapping(self, input_mapping):
```
Comment (thunterdb) [inline python/sparkdl/graph/input.py:94]: these two functions are either too much or not enough: either you should provide some tests and some doc examples, or not include them. Since they are not used elsewhere, let's put then in a separate PR for now.

File: python/tests/transformers/tf_transformer_test.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+def _build_local_features():
+    """
+    Build numpy array (i.e. local) features.
+    """
+    # Build local features and DataFrame from it
+    local_features = []
+    for idx in range(100):
+        _dict = {'idx': idx}
+        for colname, _ in _input_mapping.items():
+            _dict[colname] = np.random.randn(_tensor_size).tolist()
```
Comment (thunterdb) [inline python/tests/transformers/tf_transformer_test.py]: let's use something deterministic instead

File: python/tests/transformers/tf_transformer_test.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import tensorframes as tfs
+
+import sparkdl.graph.utils as tfx
+from sparkdl.graph.input import TFInputGraph
+from sparkdl.transformers.tf_tensor import TFTransformer
+
+from ..tests import SparkDLTestCase
+
+class TFTransformerTests(SparkDLTestCase):
+    def test_graph_novar(self):
```
Comment (thunterdb) [inline python/tests/transformers/tf_transformer_test.py:31]: great test, we will be able to add some more pretty easily after that

File: python/sparkdl/transformers/tf_tensor.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import logging
+import tensorflow as tf
+from tensorflow.python.tools import optimize_for_inference_lib as infr_opt
+import tensorframes as tfs
+
+from pyspark.ml import Transformer
+
+import sparkdl.graph.utils as tfx
+from sparkdl.graph.input import TFInputGraph
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py]: unused

File: python/sparkdl/transformers/tf_tensor.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+from sparkdl.param import (keyword_only, SparkDLTypeConverters, HasInputMapping,
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py]: `SparkDLTypeConverters` is unused

File: python/sparkdl/transformers/tf_tensor.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+logger = logging.getLogger('sparkdl')
+
+class TFTransformer(Transformer, HasTFInputGraph, HasTFHParams, HasInputMapping, HasOutputMapping):
+    """
+    Applies the TensorFlow graph to the array column in DataFrame.
+
+    Restrictions of the current API:
+
+    We assume that
+    - All graphs have a "minibatch" dimension (i.e. an unknown leading
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py]: `All the inputs of the graph`

File: python/sparkdl/transformers/tf_tensor.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+logger = logging.getLogger('sparkdl')
+
+class TFTransformer(Transformer, HasTFInputGraph, HasTFHParams, HasInputMapping, HasOutputMapping):
+    """
+    Applies the TensorFlow graph to the array column in DataFrame.
+
+    Restrictions of the current API:
+
+    We assume that
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py:38]: `The transformer is expected to work on blocks of data at the same time.`

File: python/sparkdl/transformers/tf_tensor.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    def _optimize_for_inference(self):
+        """ Optimize the graph for inference """
+        gin = self.getTFInputGraph()
+        input_mapping = self.getInputMapping()
+        output_mapping = self.getOutputMapping()
+        input_node_names = [tfx.op_name(tnsr_name) for _, tnsr_name in input_mapping]
+        output_node_names = [tfx.op_name(tnsr_name) for tnsr_name, _ in output_mapping]
+
+        # NOTE(phi-dbq): Spark DataFrame assumes float64 as default floating point type
+        opt_gdef = infr_opt.optimize_for_inference(gin.graph_def,
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py:73]: Great find! I could not find any guarantees about the stability of this function (it is part of a program). Do you know if they could become deprecated in the future?

File: python/sparkdl/transformers/tf_tensor.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                                                   input_node_names,
+                                                   output_node_names,
+                                                   tf.float64.as_datatype_enum)
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py]: With respect to your note, put a comment here, this is the place where we will eventually have to do changes.

File: python/sparkdl/transformers/tf_tensor.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            out_tnsr_op_names = [tfx.op_name(tnsr_name) for tnsr_name, _ in output_mapping]
+            tf.import_graph_def(graph_def=graph_def, name='', return_elements=out_tnsr_op_names)
+
+            feed_dict = dict((tfx.op_name(tnsr_name, graph), col_name)
+                             for col_name, tnsr_name in input_mapping)
+            fetches = [tfx.get_tensor(tnsr_op_name, graph) for tnsr_op_name in out_tnsr_op_names]
+
+            out_df = tfs.map_blocks(fetches, analyzed_df, feed_dict=feed_dict)
+
+            # We still have to rename output columns
```
Comment (thunterdb) [inline python/sparkdl/transformers/tf_tensor.py:99]: Indeed, we still need to that, annoyingly

**Review by thunterdb — CHANGES_REQUESTED** [review]
@phi-dbq this looks good. I just have a few comments that will be quick to address.

File: python/sparkdl/graph/input.py (hunk 1 above)
Comment (phi-dbq) [reply python/sparkdl/graph/input.py:94]: Well, they are actually in our API design. Let me add some tests for these guys.

File: python/sparkdl/transformers/tf_tensor.py (hunk 7 above)
Comment (phi-dbq) [reply python/sparkdl/transformers/tf_tensor.py:38]: Added this 3 lines below

File: python/sparkdl/transformers/tf_tensor.py (hunk 8 above)
Comment (phi-dbq) [reply python/sparkdl/transformers/tf_tensor.py:73]: The API looks pretty solid. I didn't see any explicit sign of this being deprecated in the near future.  
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference_lib.py#L89-L104

**Comment by phi-dbq** [comment]
@thunterdb I addressed the last round of comments. Would you like to take a look?

File: python/sparkdl/graph/input.py (hunk 1 above)
Comment (thunterdb) [reply python/sparkdl/graph/input.py:94]: Ok I see it below.

**Review by thunterdb — COMMENTED** [review]
Looks good to me.
